### PR TITLE
Adds new parameter details for the connection failed event.

### PR DESCRIPTION
**Pull request by damencho**
Adds suspend and shard_changed details.

### CODE REVIEW

File: modules/xmpp/xmpp.js [hunk 1]
```diff
@@ -509,4 +512,45 @@ export default class XMPP extends Listenable {
...
+                const header = parts.shift();
+                const value = parts.join(': ');
+
+                headers[header] = value;
+            });
+
+            /* eslint-disable camelcase */
+            details.shard_changed
+                = this.options.deploymentInfo.shard
+                    !== headers['x-jitsi-shard'];
```
Comment (bgrozev) [inline modules/xmpp/xmpp.js]: I see "X-Jitsi-Shard" coming from meet.jit.si, and [HTTP headers seem to be case-insensitive](https://tools.ietf.org/html/rfc7230#section-3.2). Is there something in strophe that converts them to all lowercase?

File: modules/xmpp/strophe.ping.js [hunk 2]
```diff
@@ -141,6 +157,52 @@ class PingConnectionPlugin extends ConnectionPlugin {
...
+     * of time.
+     */
+    wasPingSuspended() {
+        const pingIntervals = this.pingExecIntervals.slice();
+
+        // we need current time, as if ping was sent now
+        // if computer sleeps we will get correct interval after next
+        // scheduled ping, bet we sometimes need that interval before waiting
+        // for the next ping, on closing the connection on error.
+        pingIntervals.push(new Date().getTime());
```
Comment (bgrozev) [inline modules/xmpp/strophe.ping.js]: Do you intentionally not use _addPingExecutionTimestamp here? It looks like calls to wasPingSuspended can make the array grow (and _addPingExecutionTimestamp will not shrink it back, as it only removes 1 element).

File: modules/xmpp/xmpp.js [hunk 3]
```diff
@@ -509,4 +512,45 @@ export default class XMPP extends Listenable {
...
+            /* eslint-enable camelcase */
+        }
+
+        // check for possible suspend
```
Comment (bgrozev) [inline modules/xmpp/xmpp.js]: I thought we also wanted the suspend duration somewhere

File: modules/xmpp/strophe.ping.js [hunk 4]
```diff
@@ -23,6 +23,19 @@ const PING_TIMEOUT = 15000;
...
+ * @type {number} number of timestamps.
+ */
+const PING_TIMESTAMPS_TO_KEEP = 12;
+
+/**
+ * If execution of pings is higher than this threshold, we consider that the
+ * ping was suspended.
+ * @type {number}
+ */
+const SUSPEND_INTERVAL_THRESHOLD = PING_TIMEOUT;
```
Comment (bgrozev) [inline modules/xmpp/strophe.ping.js]: What is the relationship between SUSPEND_INTERVAL_THRESHOLD and PING_TIMEOUT? 

Wouldn't we want SUSPEND_INTERVAL_THRESHOLD to depend on PING_INTERVAL instead (if PING_INTERVAL is 1000, we could detect suspension sooner; if it is 100_000 we would have false positive detections)?

File: JitsiConnection.js [hunk 5]
```diff
@@ -23,11 +23,13 @@ export default function JitsiConnection(appID, token, options) {
     this.options = options;
     this.xmpp = new XMPP(options, token);
 
+    /* eslint-disable max-params */
     this.addEventListener(JitsiConnectionEvents.CONNECTION_FAILED,
-        (errType, msg) => {
+        (errType, msg, credentials, details) => {
```
Comment (bgrozev) [inline JitsiConnection.js]: We explicitly disable the max-params rule in order to add an extra parameter which is always "undefined" and always ignored :)

Comment (bgrozev) [reply JitsiConnection.js]: Ah, I see now that it is used elsewhere.

File: modules/xmpp/xmpp.js (hunk 1 above)
Comment (damencho) [reply modules/xmpp/xmpp.js]: Hum, I suppose the XHR API works that way. All strophe does is: req.xhr.getAllResponseHeaders();
https://github.com/strophe/strophejs/commit/93d63d3f7adc1ea126e6ed57f1b5fabf6a34c765#diff-c8241e10984282f97b0a3814d0bb41e7R619
And I've been testing this with meet.jit.si.

File: modules/xmpp/strophe.ping.js (hunk 2 above)
Comment (damencho) [reply modules/xmpp/strophe.ping.js]: Well my thinking was working on a copy of the array and compute it as if a ping was just sent it, does it make sense. It is a copy of the array, so nothing grows.

File: modules/xmpp/xmpp.js (hunk 3 above)
Comment (damencho) [reply modules/xmpp/xmpp.js]: Ok, I will modify it and add that.

File: modules/xmpp/strophe.ping.js (hunk 4 above)
Comment (damencho) [reply modules/xmpp/strophe.ping.js]: Yep, I was thinking to make them dependent ... but if we are always sending the interval maybe I will remove this threshold and we detect this on the analytics side of things and move the threshold there.

**Comment by damencho** [comment]
@bgrozev I pushed the changes and tests finally passed. Does it look ok now?

File: modules/xmpp/xmpp.js [hunk 6]
```diff
@@ -509,4 +512,45 @@ export default class XMPP extends Listenable {
...
+        details.suspendTime = this.connection.ping.getPingSuspendTime();
```
Comment (bgrozev) [inline modules/xmpp/xmpp.js]: Please use "suspend_time" if this goes to analytics

**Pull request merged**